nav2_map_server/src/vo_server/vector_object_server.cpp

```diff
@@ -0,0 +1,565 @@
+        response->success = false;
+      }
+    }
+  }
+
+  // Process circles
+  for (auto req_crcl : request->circles) {
+    nav2_msgs::msg::CircleObject::SharedPtr new_params =
+      std::make_shared<nav2_msgs::msg::CircleObject>(req_crcl);
+
+    auto shape = shapes_.find(unparseUUID(new_params->uuid.uuid.data()));
+    if (shape != shapes_.end()) {
+      // Vector object with given UUID was found: updating it
+      // Check that found shape has correct type
+      if (shape->second->getType() != CIRCLE) {
+        RCLCPP_ERROR(
+          get_logger(),
+          "Shape (UUID: %s) is not a circle type",
+          shape->second->getUUID().c_str());
+        response->success = false;
+        // Do not add this shape
+        continue;
+      }
+
+      std::shared_ptr<Circle> circle = std::static_pointer_cast<Circle>(shape->second);
+
+      // Preserving old parameters for the case, if new ones to be incorrect
+      nav2_msgs::msg::CircleObject::SharedPtr old_params = circle->getParams();
+      if (!circle->setParams(new_params)) {
+        // Restore old parameters
+        circle->setParams(old_params);
+        // ... and set the failure to return
+        response->success = false;
+      }
+    } else {
+      // Vector Object with given UUID was not found: creating a new one
+      std::shared_ptr<Circle> circle = std::make_shared<Circle>(node);
+      if (circle->setParams(new_params)) {
+        shapes_.insert({circle->getUUID(), circle});
+      } else {
+        response->success = false;
+      }
+    }
+  }
+
+  switchMapUpdate();
+}
+
+void VectorObjectServer::getShapesCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::GetShapes::Request>/*request*/,
+  std::shared_ptr<nav2_msgs::srv::GetShapes::Response> response)
+{
+  std::shared_ptr<Polygon> polygon;
+  std::shared_ptr<Circle> circle;
+
+  for (auto shape : shapes_) {
+    switch (shape.second->getType()) {
+      case POLYGON:
+        polygon = std::static_pointer_cast<Polygon>(shape.second);
+        response->polygons.push_back(*(polygon->getParams()));
+        break;
+      case CIRCLE:
+        circle = std::static_pointer_cast<Circle>(shape.second);
+        response->circles.push_back(*(circle->getParams()));
+        break;
+      default:
+        RCLCPP_WARN(get_logger(), "Unknown shape type (UUID: %s)", shape.second->getUUID().c_str());
+    }
+  }
+}
+
+void VectorObjectServer::removeShapesCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::RemoveShapes::Request> request,
+  std::shared_ptr<nav2_msgs::srv::RemoveShapes::Response> response)
+{
+  // Initialize result with true. If one of the required vector object was not found,
+  // set it to false.
+  response->success = true;
+
+  if (request->all_objects) {
+    // Clear all objects
+    for (auto shape : shapes_) {
+      shape.second.reset();
+    }
+    shapes_.clear();
```

Comment:
clear() should reset all the shared pointers, no?

Reply:
Yep, this is not needed: at `clear()` stage, all unused smart pointers inside to be destroyed. Fixed in https://github.com/ros-planning/navigation2/pull/3930/commits/1bb10dcbb2b00962f896bd46b81d1a9070846bbb